# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
#datapath = 
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")

data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*Replace the text in this markdown cell with your predictions, and be sure to provide justification for your guess.*

In [3]:
#I think the Logistic Regression will peform better because the nature of the question is binary and that this is the fastest most efficent method.

## Split the Data into Training and Testing Sets

In [4]:
# Create the labels set `y` and features DataFrame `X`

#becasue I'm looking to predict the value for the spam column, this is my target and will be my y value
#I drop the spam column from my features to create the X data
y = data["spam"]
X = data.copy()
X = data.drop(columns="spam")


In [6]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [7]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y)

In [8]:
#inspect the df
X_train.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
4038            0.00               0.00           0.00           0.0   
4275            0.00               0.36           0.00           0.0   
193             0.00               0.29           0.87           0.0   
142             0.50               0.46           0.34           0.0   
4332            0.26               0.00           0.26           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
4038           0.00            0.00              0.00                0.00   
4275           0.00            0.00              0.00                0.00   
193            0.29            0.00              0.14                0.00   
142            0.15            0.03              0.00                0.19   
4332           0.52            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
4038             0.00            0.00  ...                   0.0        0.000   
4275             0.00            0.73  ...                   0.0        3.838   
193              0.00            0.43  ...                   0.0        0.000   
142              0.57            0.65  ...                   0.0        0.011   
4332             0.00            0.26  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
4038        0.000          0.0        0.000        0.000        0.000   
4275        0.000          0.0        0.062        0.000        0.503   
193         0.000          0.0        0.585        0.046        0.046   
142         0.077          0.0        0.335        1.281        0.125   
4332        0.000          0.0        0.038        0.000        0.038   

      capital_run_length_average  capital_run_length_longest  \
4038                       1.947                          12   
4275                       6.912                          43   
193                        5.020                         132   
142                        7.202                         595   
4332                       1.541                          12   

      capital_run_length_total  
4038                        37  
4275                       394  
193                        979  
142                       2413  
4332                       202  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [23]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instances
scaler = StandardScaler()

In [24]:
# Fit the Standard Scaler with the training data
X_scaler = scaler.fit(X_train)

In [25]:
# Scale the training data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)


## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [12]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

lr_model = LogisticRegression(random_state=1)
lr_model.fit(X_train_scaled,y_train)
print(f"Training Data Score:  {lr_model.score(X_train_scaled,y_train)}")
print(f"Testing Data Score:  {lr_model.score(X_test_scaled,y_test)}")

Training Data Score:  0.9255072463768116
Testing Data Score:  0.9287576020851434


In [13]:
# Make and save testing predictions with the saved logistic regression model using the test data
predictions = lr_model.predict(X_test_scaled)
# Review the predictions
predictions

array([0, 0, 0, ..., 0, 1, 0])

In [14]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test,predictions)

0.9287576020851434

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [26]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
random_model = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train_scaled,y_train)

print(f'Training Score: {random_model.score(X_train_scaled, y_train)}')
print(f'Testing Score: {random_model.score(X_test_scaled, y_test)}')

Training Score: 0.9994202898550725
Testing Score: 0.9600347523892268


In [27]:
# Make and save testing predictions with the saved logistic regression model using the test data
predictions2 = random_model.predict(X_test_scaled)

# Review the predictions
predictions2

array([0, 0, 0, ..., 0, 1, 0])

In [28]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
acc_score = accuracy_score(y_test,predictions2)
acc_score

0.9600347523892268

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

#The random forest model performed better for the analysis - yielding a 96% compared to a 92.8% for the logistic regression.  My prediction was incorrect as the Random Forest model was able to group and define more like elements within the data set to better predict where the mail item should go - spam or inbox.